LION/optimizers/ARsolver.py

```diff
@@ -0,0 +1,283 @@
+# This file is part of LION library
+# License : BSD-3
+#
+# Author  : Zakhar Shumaylow, Charlie Shoebridge
+# Modifications: Ander Biguri
+# =============================================================================
+
+import numpy as np
+import torch
+from torch.optim.optimizer import Optimizer
+from tqdm import tqdm
+from LION.CTtools.ct_geometry import Geometry
+from LION.classical_algorithms.fdk import fdk
+from tomosipo.torch_support import to_autograd
+from LION.exceptions.exceptions import LIONSolverException
+from LION.models.LIONmodel import LIONmodel, ModelInputType
+from LION.optimizers.LIONsolver import LIONsolver, SolverParams
+from LION.utils.parameter import LIONParameter
+
+
+class ARParams(SolverParams):
+    def __init__(
+        self,
+        optimizer_params: LIONParameter,
+        early_stopping: bool = False,
+        no_steps: int = 150,
+        step_size: float = 1e-6,
+        beta_rate: float = 0.95,
+        lambd_gp: float = 1e-1,
+    ):
+        super().__init__()
+        self.early_stopping = early_stopping
+        self.no_steps = no_steps
+        self.step_size = step_size
+        self.optimizer_params = optimizer_params
+        self.beta_rate = beta_rate
+        self.lambd_gp = lambd_gp
+
+
+class ARSolver(LIONsolver):
+    def __init__(
+        self,
+        model: LIONmodel,
+        optimizer: Optimizer,
+        recon_optimizer_type,
+        geometry: Geometry,
+        verbose: bool = True,
+        device: torch.device = torch.device(f"cuda:{torch.cuda.current_device()}"),
+        solver_params: ARParams | None = None,
+    ) -> None:
+        super().__init__(
+            model, optimizer, self.wgan_loss, geometry, verbose, device, solver_params
+        )
+        self.recon_optimizer_type = recon_optimizer_type
+        self.A = to_autograd(self.op, num_extra_dims=1)
+        self.AT = to_autograd(self.op.T, num_extra_dims=1)
+        self.alpha = 1.0
+
+    def estimate_alpha(self, dataloader):
+        if dataloader is not None:
+            residual = 0.0
+            for _, (sino, target) in enumerate(dataloader):
+                residual += torch.linalg.vector_norm(
+                    self.AT(self.A(target) - sino), dim=(2, 3)
+                ).mean()
+                # residual += torch.sqrt(((self.AT(self.A(target) - data))**2).sum())
+            self.alpha = residual / len(dataloader.dataset)
+        if self.verbose:
+            print("Estimated alpha: " + str(self.alpha))
```

Comment:
@Zakobian is this the power method for the operator norm? or something else?

Reply:
Nope, this is estimating the regularisation parameter